src/pocketmine/Player.php

```diff
@@ -2788,10 +2791,27 @@ public function handlePlayerAction(PlayerActionPacket $packet) : bool{
 				//TODO: destroy-progress level event
 				break;
 			case PlayerActionPacket::ACTION_START_SWIMMING:
-				break; //TODO
+				$block = $this->level->getBlock($this);
+
+				$ev = new PlayerToggleSwimEvent($this, true);
+				$this->server->getPluginManager()->callEvent($ev);
+				if(!$block instanceof Water or $ev->isCancelled()){
```

Comment:
I think it's better to cancel the event before calling it if the player is not in water, to give plugins a chance to alter the outcome.